heft-plugins/heft-jest-plugin/src/transformers/BuildTransformer.ts

```diff
@@ -0,0 +1,449 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import { Path, FileSystem, FileSystemStats, JsonObject } from '@rushstack/node-core-library';
+import createCacheKeyFunction from '@jest/create-cache-key-function';
+import type { AsyncTransformer, SyncTransformer, TransformedSource, TransformOptions } from '@jest/transform';
+import type { Config } from '@jest/types';
+
+import { HeftJestDataFile, IHeftJestDataFileJson } from '../HeftJestDataFile';
+
+// This caches heft-jest-data.json file contents.
+// Map from jestOptions.rootDir --> IHeftJestDataFileJson
+const dataFileJsonCache: Map<string, IHeftJestDataFileJson> = new Map();
+
+// Synchronous delay that doesn't burn CPU cycles
+function delayMs(milliseconds: number): void {
+  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, milliseconds);
+}
+
+async function delayMsAsync(milliseconds: number): Promise<void> {
+  await new Promise<void>((resolve: () => void): void => {
+    setTimeout(resolve, milliseconds);
+  });
+}
+
+const DEBUG_TRANSFORM: boolean = false;
+
+// Tolerate this much inaccuracy in the filesystem time stamps
+const TIMESTAMP_TOLERANCE_MS: number = 15;
+
+// Wait this long after a .js file's timestamp changes before starting to read it; this gives time
+// for the contents to get flushed to disk.
+const FLUSH_TIME_MS: number = 500;
+
+// Wait this long for the .js file to be written before giving up.
+const MAX_WAIT_MS: number = 7000;
+
+// Shamefully sleep this long to avoid consuming CPU cycles
+const POLLING_INTERVAL_MS: number = 50;
+
+// See: https://github.com/facebook/jest/blob/e0b33b74b5afd738edc183858b5c34053cfc26dd/packages/jest-create-cache-key-function/src/index.ts#L35
+type NewCacheKeyOptions = Pick<TransformOptions, 'config' | 'configString' | 'instrument'>;
+type NewGetCacheKeyFunction = (
+  sourceText: string,
+  sourcePath: Config.Path,
+  options: NewCacheKeyOptions
+) => string;
+
+/**
+ * This Jest transformer maps TS files under a 'src' folder to their compiled equivalent under 'lib'.
+ *
+ * @privateRemarks
+ * Implements SyncTransformer interface instead of AsyncTransformer since async is only supported
+ * in ESM, which is still considered experimental:
+ * https://github.com/facebook/jest/issues/11226#issuecomment-804449688
+ */
+export class BuildTransformer implements AsyncTransformer, SyncTransformer {
+  /**
+   * Read heft-jest-data.json, which is created by the JestPlugin.  It tells us
+   * which emitted output folder to use for Jest.
+   */
+  private static _getHeftJestDataFileJson(rootDir: string): IHeftJestDataFileJson {
```

Comment:
Why a blend of free functions and private static functions? The former are better for if we ever run this through a minifier; some folks prefer the latter for casting and mocking.

Reply:
Sticking with the statics. I like having the class be able to implement the correct types.